Add Kafka producer

MatheusMiranda opened the pull request:
Signed-off-by: Matheus Miranda <matheusmirandalacerda@gmail.com>
Signed-off-by: KahCosta <kamillacosta.unb@gmail.com>

**Proposed changes**:
- Added producer to stream events to Kafka Broker. Can be used with both PLAINTEXT and SSL protocols.
- Generalized broker instantiation at 'run.py'.
- Created 'event brokers' directory and split producers classes into different files.
- Added documentation to Kafka event broker.

Related to issue: #1553 
- ...

**Status (please check what you already did)**:
- [X] made PR ready for code review
- [X] added some tests for the functionality
- [X] updated the documentation
- [ ] updated the changelog

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/RasaHQ/rasa_core?pullRequest=1648) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/RasaHQ/rasa_core?pullRequest=1648) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/RasaHQ/rasa_core?pullRequest=1648) it.</sub>

MatheusMiranda commented:
Nice, thank you!

MatheusMiranda commented:
Hey, @tmbo

Build is breaking on "Test docs" stage for this PR. I've tried to find the reason, but i was not able to reproduce the error.

Any idea on what it could be?

ExarchD commented:
Is there any update on this pull request?

MatheusMiranda commented:
> Is there any update on this pull request?

Build Passing now. It was only a missing dependency on setup.

MatheusMiranda commented:
Nice. 
I'm going to update it, and rebase the branch with master.

> This is a great addition and I'd love to merge this. Because of the added kafka broker and an additional broker we already added, we changed the way how the user specifies which broker to use.
> 
> We added a `type` field to the `event_broker` configuration which selects the broker to instantiate, e.g.
> 
> ```
> event_broker:
>   type: file
>   path: rasa_event.log
> ```
> This will instantiate the `FileProducer` instead of the pika producer.
> 
> To merge your PR, it would be great if you could
> 
> * merge or rebase onto `master`
> * add the Kafka broker to `from_endpoint_config` in `broker.py`
> * remove the additional `--event_broker` argument to run
> 
> Would it be possible for you to make these changes? After that I'd do a full code review.

tmbo commented on docs/brokers.rst:
```diff
@@ -96,5 +105,131 @@ example:
+
+    python3 -m rasa_core.run -d models/dialogue -u models/nlu/current --event_broker kafka_producer --endpoints endpoints.yml
+
+Using SASL_PLAINTEXT protocol the endpoints file must have the following entries:
```
```suggestion
Using ``SASL_PLAINTEXT`` protocol the endpoints file must have the following entries:
```

tmbo commented on docs/brokers.rst:
```diff
@@ -96,5 +105,131 @@ example:
+Authentication and authorization
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Rasa Core's Kafka producer accepts two types of security protocols - SASL_PLAINTEXT and SSL.
```
```suggestion
Rasa Core's Kafka producer accepts two types of security protocols - ``SASL_PLAINTEXT`` and ``SSL``.
```

tmbo commented on docs/brokers.rst:
```diff
@@ -96,5 +105,131 @@ example:
+Rasa Core's Kafka producer accepts two types of security protocols - SASL_PLAINTEXT and SSL.
+
+For development environment, or if the brokers servers and clients are located
+into the same machine, you can use simple authentication with SASL_PLAINTEXT.
```
```suggestion
into the same machine, you can use simple authentication with ``SASL_PLAINTEXT``.
```

tmbo commented on docs/brokers.rst:
```diff
@@ -96,5 +105,131 @@ example:
+By using this protocol, the credentials and messages exchanged between the clients and servers
+will be sent in plaintext. Thus, this is not the most secure approach, but since it's simple
+to configure, it is useful for simple cluster configurations.
+SASL_PLAINTEXT protocol requires the setup of the 'username' and 'password'
```
```suggestion
``SASL_PLAINTEXT`` protocol requires the setup of the ``username`` and ``password``
```

tmbo commented on docs/brokers.rst:
```diff
@@ -96,5 +105,131 @@ example:
+will be sent in plaintext. Thus, this is not the most secure approach, but since it's simple
+to configure, it is useful for simple cluster configurations.
+SASL_PLAINTEXT protocol requires the setup of the 'username' and 'password'
+previously configured in the broker server.
```
what do you mean here, where is this configured? I guess in the running Kafka instanc, right?

MatheusMiranda replied:
It's configured on both client and server. There is communication between client<->server and between servers itself on cluster. The security protocol being used may differ for these two communications.

This is referred to producer(client) <-> server communication.

tmbo commented on rasa_core/broker.py:
```diff
@@ -132,3 +136,62 @@ def publish(self, event: Dict) -> None:
+
+
+class KafkaProducer(EventChannel):
+    def __init__(self, host, sasl_plain_username=None,
```
mhm why not just `sasl_username`? (or is it called `sasl_plain_username` in kafka?)

MatheusMiranda replied:
That's because it's called `sasl_plain_protocol` in `kafka-python` lib. But you're right, it's clear by the security_protocol.

tmbo commented on rasa_core/broker.py:
```diff
@@ -132,3 +136,62 @@ def publish(self, event: Dict) -> None:
+
+class KafkaProducer(EventChannel):
+    def __init__(self, host, sasl_plain_username=None,
+                 sasl_plain_password=None, ssl_cafile=None,
```
```suggestion
                 sasl_plain_password=None, 
                 ssl_cafile=None,
```

tmbo commented on rasa_core/broker.py:
```diff
@@ -132,3 +136,62 @@ def publish(self, event: Dict) -> None:
+class KafkaProducer(EventChannel):
+    def __init__(self, host, sasl_plain_username=None,
+                 sasl_plain_password=None, ssl_cafile=None,
+                 ssl_certfile=None, ssl_keyfile=None,
```
```suggestion
                 ssl_certfile=None, 
                 ssl_keyfile=None,
```

tmbo commented on rasa_core/broker.py:
```diff
@@ -132,3 +136,62 @@ def publish(self, event: Dict) -> None:
+
+
+class KafkaProducer(EventChannel):
+    def __init__(self, host, sasl_plain_username=None,
```
```suggestion
    def __init__(self, host, 
                 sasl_plain_username=None,
```

tmbo commented on rasa_core/event_brokers/__init__.py:
```diff
@@ -0,0 +1,10 @@
+
+event_broker_classes = [
+    PikaProducer, KafkaProducer
+]  # type: List[InputChannel]
```
these types need to be imported at the top:
```python
import typing
from typing import List, Text, Dict

if typing.TYPE_CHECKING:
    from rasa_core.channels import InputChannel
```

tmbo commented on rasa_core/run.py:
```diff
@@ -72,6 +72,17 @@ def _create_single_channel(channel, credentials):
                 "is a proper name of a class in a module.".format(channel))
 
 
+def create_event_broker(event_broker=None, event_broker_config=None):
```
this function is not used

tmbo commented on rasa_core/event_brokers/kafka_producer.py:
```diff
@@ -0,0 +1,63 @@
+import logging
```
this code is duplicated, let's keep the code in `broker.py` for now

MatheusMiranda replied:
I've forgot to delete these files haha

tmbo commented on rasa_core/event_brokers/pika_producer.py:
```diff
@@ -0,0 +1,44 @@
+import logging
```
same here

tmbo commented on setup.py:
```diff
@@ -67,7 +67,8 @@
     "keras-applications==1.0.6",
-    "keras-preprocessing==1.0.5"
+    "keras-preprocessing==1.0.5",
+    "kafka-python~=1.4.4"
```
```suggestion
    "kafka-python~=1.4"
```

tmbo commented:
Alright, sounds great - please let me know when this is ready for another review :Rocket:

MatheusMiranda commented:
> Alright, sounds great - please let me know when this is ready for another review

I think now it's ready for review ;):+1:

tmbo commented on rasa_core/run.py:
```diff
@@ -183,7 +183,6 @@ def load_agent(core_model, interpreter, endpoints,
     logging.getLogger('engineio').setLevel(logging.WARN)
     logging.getLogger('matplotlib').setLevel(logging.WARN)
     logging.getLogger('socketio').setLevel(logging.ERROR)
-    logging.getLogger('pika').setLevel(logging.ERROR)
```
is there a reason to remove this logger configuration?

tmbo replied:
nevermind makes sense with the change to the pike init :+1:

MatheusMiranda commented:
> alright looks great - awesome job 
> 
> One quick question about a removed line, but other than that we are ready to go.

Nice, thanks for the review! :grin: :v:
